### PR TITLE
#160577483 Liking/Disliking Articles Functionality 

### PR DESCRIPTION
**Pull request by IncredibleBlackMan**
#### What does this PR do?
Make user's of Author's Haven able to like/dislike and delete a like/dislike to an article.
#### Description of Task to be completed?
Have the following endpoints working
```POST /api/articles/<slug>/like/``` - post a like/dislike
```DELETE /api/articles/<slug>/like/``` - delete a like/dislike
#### How should this be manually tested?
- Clone [this](https://github.com/andela/ah-shakas.git) repository and cd into the folder.
- Run the command ```git checkout ft-liking-unliking-articles-160577483```
- Create a ```.env``` file with the following values
```
DB_NAME=
DB_HOST=
DB_PASSWORD=
DB_USER=
EMAIL_HOST=smtp.sendgrid.net
EMAIL_HOST_USER=
EMAIL_HOST_PASSWORD=
EMAIL_PORT=
```
- Install the project dependencies by running ```pipenv install```
- Then to run the tests ren the command ```python manage.py test```
- Using [Postman](https://www.getpostman.com/) test every endpoint above after running the command ```python manage.py runserver```
- Create a user account by accessing ```POST api/users/``` and sending the following payload ```{ "user":{ "username":"username", "email":"username@gmail.com", "password": "password1234"}}```
- Confirm signup from the email sent to the address posted in the payload.
- Then login to get token to perform manual tests
_key_: Content-Type value: application/json
_key_: Authorization value: JWT Token generated either on signup or login
- Create an article by accessing ```POST api/articles/``` and pass in the following payload ```{"article":{"title":"title", "Title", "body": "My article"}}```
- Test the endpoins above by passing the following payload ```{"likes":true}```
#### Any background context you want to provide?
Author's Haven application has three article states when it comes to likes. A reader can have no response to an article, and either likes or dislike an article. One can exist in only one of these states.
#### What are the relevant pivotal tracker stories?
[#160577483 Users should be able to like and or dislike articles](https://www.pivotaltracker.com/n/projects/2198533)
#### Screenshots (if appropriate)
screen shot 2018-10-16 at 11 07 31
screen shot 2018-10-16 at 11 07 57

screen shot 2018-10-16 at 11 09 06

screen shot 2018-10-16 at 11 09 39

screen shot 2018-10-16 at 11 10 10

### CODE REVIEW

File: authors/apps/articles/models.py [hunk 1]
```diff
@@ -14,6 +14,9 @@ class ArticlesModel(models.Model):
     created_at = models.DateTimeField(auto_now_add=True, auto_now=False)
     updated_at = models.DateTimeField(auto_now=True, auto_now_add=False)
     author = models.ForeignKey(User, related_name='article', on_delete=models.CASCADE)
+    likes = models.ManyToManyField(User, blank=True, related_name='LikesUnlikesModel.user+')
```
Comment (lukorito) [inline authors/apps/articles/models.py]: Why the related name with a dot? I get the + part (You dont want a reverse relationship). I dont get the other part

File: authors/apps/articles/serializers.py [hunk 2]
```diff
@@ -93,3 +106,16 @@ def to_representation(self,instance):
...
+
+class LikesDislikesSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = LikesDislikes
+        fields = '__all__'
+        validators = [
+            UniqueTogetherValidator(
+                queryset=LikesDislikes.objects.all(),
+                fields = ('article', 'reader'),
+                message = 'You have already likesd this article.'
```
Comment (lukorito) [inline authors/apps/articles/serializers.py]: Check typo on `likesd`

**Review by lukorito — CHANGES_REQUESTED** [review]

**Comment by lukorito** [comment]
Some great work on this one @Ramon3162

File: authors/apps/articles/models.py (hunk 1 above)
Comment (IncredibleBlackMan) [reply authors/apps/articles/models.py]: Hey @lukorito, that's just naming to make it more descriptive. The code still works just fine without it, and with the + part alone, and yes, I do not want a reverse relationship from the User model to this field.

File: authors/apps/articles/serializers.py (hunk 2 above)
Comment (IncredibleBlackMan) [reply authors/apps/articles/serializers.py]: Thanks for this observation, on it.

**Review by lukorito — APPROVED** [review]

**Comment by lukorito** [comment]
I have tested this and it is working. It should be merged

File: authors/apps/articles/serializers.py [hunk 3]
```diff
@@ -1,14 +1,15 @@
 from rest_framework import serializers
 from django.db.models import Avg
 from django.core.validators import MinValueValidator, MaxValueValidator
+from rest_framework.validators import UniqueTogetherValidator
```
Comment (lukorito) [inline authors/apps/articles/serializers.py]: @Ramon3162 This file has been imported on line 10

File: authors/apps/articles/tests/test_likes.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            "user": {
+                "username": "seconduser",
+                "password": "password134",
+                "email": "pass@g.com"
+            }
+        }
+        
+        self.author_token = self.create_and_login_user()
+        self.article_slug = self.create_article(token=self.author_token)
+        # self.user_token = self.create_and_login_user(second_user)
```
Comment (codingedward) [inline authors/apps/articles/tests/test_likes.py]: Hello @Ramon3162, let's get rid of this commented code.

File: authors/apps/articles/views.py [hunk 5]
```diff
@@ -342,3 +342,174 @@ def delete(self, request, slug):
         data = {"article":article_serializer.data}
         data["message"] = "unfavourited"
         return Response(data, status.HTTP_200_OK)
+
```
Comment (lukorito) [inline authors/apps/articles/views.py:362]: @Ramon3162 Fix the spacing here. Just add an extra line

File: authors/apps/articles/views.py [hunk 6]
```diff
@@ -7,12 +7,12 @@
...
 from .permissions import IsOwnerOrReadonly
-from .models import ArticlesModel, Comment, Rating, Favourite
+from .models import ArticlesModel, Comment, Rating, Favourite, LikesDislikes
 from .serializers import (ArticlesSerializers, 
         CommentsSerializers, RatingSerializer,
-        FavouriteSerializer)
+        FavouriteSerializer, LikesDislikesSerializer)
 from .renderers import ArticlesRenderer, RatingJSONRenderer, FavouriteJSONRenderer
-from .permissions import IsOwnerOrReadonly
+
```
Comment (lukorito) [inline authors/apps/articles/views.py:23]: It should be one line here.

**Review by lukorito — CHANGES_REQUESTED** [review]

File: authors/apps/articles/tests/test_likes.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+        self.slug = self.create_article()
+        self.article_url = api_reverse('articles:article-details', {self.slug: 'slug'})
+        self.like_url = api_reverse('articles:article-like', {self.slug: 'slug'})
+
+    def create_like(self, like=None):
+        if not like:
+            like=self.like
```
Comment (codingedward) [inline authors/apps/articles/tests/test_likes.py]: We could have this as:
```python
like = self.like
```
Note the spacing, in PEP8, we only leave out spaces when we are passing keyword arguments or when providing a default value in a method's signature.

File: authors/apps/articles/tests/test_likes.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            like, 
+            'json', 
+            HTTP_AUTHORIZATION=self.create_and_login_user(self.second_user)
+        )
+
+        return response
+
+    def create_dislike(self, like=None):
+        if not like:
+            like=self.dislike
```
Comment (codingedward) [inline authors/apps/articles/tests/test_likes.py]: Ditto

File: authors/apps/articles/serializers.py (hunk 3 above)
Comment (IncredibleBlackMan) [reply authors/apps/articles/serializers.py]: Thanks for that catch brotherman, on it. Won't happen again.

**Review by codingedward — COMMENTED** [review]
Hi @Ramon3162, I have reviewed your code and highlighted a few changes we might want to address.

Lastly, I really like the way you have refactored your tests as they are pretty DRY 🎊

File: authors/apps/articles/tests/test_likes.py (hunk 4 above)
Comment (IncredibleBlackMan) [reply authors/apps/articles/tests/test_likes.py]: Making the change right away. Thanks for noticing this.

File: authors/apps/articles/views.py (hunk 5 above)
Comment (IncredibleBlackMan) [reply authors/apps/articles/views.py:362]: Done

**Review by lukorito — APPROVED** [review]

File: authors/apps/articles/serializers.py (hunk 3 above)
Comment (IncredibleBlackMan) [reply authors/apps/articles/serializers.py]: Noted, I have made the change already.

File: authors/apps/articles/views.py (hunk 5 above)
Comment (IncredibleBlackMan) [reply authors/apps/articles/views.py:362]: Done.

**Comment by lukorito** [comment]
@Ramon3162 Great work. You may want to add titles to your screenshots

**Pull request merged**